File: js/src/days/day06.js
import { mult } from "../lib/iterators.js"
import { readNumberList } from "../lib/parsing.js"
import * as runner from "../lib/runner.js"

console.log("Solving AoC 2023 day 06")

const parseInput = (input) => {
  const lines = input.split('\n')
  const times = readNumberList(lines[0])
  const distances = readNumberList(lines[1])
  return {times, distances}
}

const part1 = (input) => {
  const {times, distances} = parseInput(input)
  const numWaysToWinList = times.map((time, i) => {
    const targetDistance = distances[i]
    const halfTime = time / 2
    const sqrtComponent = Math.sqrt(halfTime**2 - targetDistance)
    const minPush = halfTime - sqrtComponent
    let floorMin = Math.floor(minPush)
    const maxPush = halfTime + sqrtComponent
    let ceilMax = Math.ceil(maxPush)
    let numWays = time - 1 // the last time push value will never win
    numWays -= floorMin // subtract the low push values that wont win
    numWays -= (time - ceilMax) // subtract the high push values that wont win
    return numWays
  })
  return mult(numWaysToWinList)
}

// const part2 = (input) => {
//   return 0
// }

await runner.testOutput('day06/example', '1', part1)
// await runner.printOutput('day06/test', part1)
// await runner.copyOutput('day06/test', part1)
// await runner.writeOutput('day06/test', '1', part1)
await runner.testOutput('day06/test', '1', part1)

// await runner.testOutput('day06/example', '2', part2)
// await runner.printOutput('day06/test', part2)
// await runner.copyOutput('day06/test', part2)
// await runner.writeOutput('day06/test', '2', part2)
// await runner.testOutput('day06/test', '2', part2)

File: swift/Sources/days/day06.swift
import Foundation

private func readNumberList(from line: Substring) -> [Int] {
    line.split(separator: " ").compactMap { Int($0) }
}

private func parseInput(_ input: String) -> (times: [Int], distances: [Int]) {
    let lines = input.split(separator: "\n")
    let timesLine = lines[0].split(separator: ":")[1]
    let distancesLine = lines[1].split(separator: ":")[1]
    let times = readNumberList(from: timesLine)
    let distances = readNumberList(from: distancesLine)
    return (times, distances)
}

// Function to calculate the number of ways to win based on given times and distances
private func solveProblem(_ input: String) -> Int {
    let (times, distances) = parseInput(input)
    let numWaysToWinList = zip(times, distances).map { time, distance -> Int in
        let halfTime = Double(time) / 2.0
        let sqrtComponent = sqrt(halfTime * halfTime - Double(distance))
        let minPush = halfTime - sqrtComponent
        let maxPush = halfTime + sqrtComponent
        let floorMin = Int(floor(minPush))
        let ceilMax = Int(ceil(maxPush))
        var numWays = time - 1 // the last time push value will never win
        numWays -= floorMin // subtract the low push values that won't win
        numWays -= (time - ceilMax) // subtract the high push values that won't win
        return numWays
    }
    return numWaysToWinList.reduce(1, *)
}

private func part1(_ input: String) -> String {
    let result = solveProblem(input)
    return String(result)
}

private func part2(_ input: String) -> String {
    let fixedInput = input.replacingOccurrences(of: " ", with: "")
    let result = solveProblem(fixedInput)
    return String(result)
}

func day06(dayKey: String) throws {
    try testAocOutput(dayKey: dayKey, inputName: "example", partKey: "1", partSolver: part1)
    try testAocOutput(dayKey: dayKey, inputName: "test", partKey: "1", partSolver: part1)

    try testAocOutput(dayKey: dayKey, inputName: "example", partKey: "2", partSolver: part2)
    try testAocOutput(dayKey: dayKey, inputName: "test", partKey: "2", partSolver: part2)
}
